### ISSUE TITLE
CLIP loads OpenAI weights without QuickGELU: embedding fidelity degraded across the whole pipeline

### ISSUE THREAD

**Issue by yves-vogl** [PixelCurator/pixelcurator#38]
## Observation

Every CLIP call site (`01_embed.py`, `03_classify_cluster.py`, `06_taxonomy.py`, `tests/test_ml_smoke.py`) does:

```python
open_clip.create_model_and_transforms("ViT-B-32", pretrained="openai")
```

open_clip warns at every model load (observed locally with the exact pinned wheels from requirements.txt, open_clip_torch >= 3.3):

```
UserWarning: QuickGELU mismatch between final model config (quick_gelu=False)
and pretrained tag 'openai' (quick_gelu=True).
```

Per the open_clip README (mlfoundations/open_clip, retrieved 2026-08-27): OpenAI pretrained weights were trained with the QuickGELU activation and should be loaded via model definitions with the `-quickgelu` postfix (`ViT-B-32-quickgelu`); using the plain variant with OpenAI weights causes an accuracy drop.

So the entire embedding space (81.5k photos), the zero-shot classification in 06_taxonomy, and the retrain classifier all run on activations that do not match the weights. It works (the ML smoke test's coarse red-vs-blue check passes with >0.9 confidence), but fidelity is measurably below what the same weights can deliver.

## Why this is not a quick mechanical fix

Switching the model tag changes the embedding space: `embeddings/clip_vitb32.npy` (~167 MB, 81,559 photos), all cosine similarities, the retrain classifier and its thresholds (CONF_MAIN=0.80 / CONF_UNSURE=0.50, tuned against the current space) are all invalidated together. Needs a coordinated migration:

1. Switch all four call sites to `ViT-B-32-quickgelu` in one change (a partial switch would mix incompatible spaces).
2. Full re-embed (~42 min at the current 32 img/s baseline, ~11 min if the parallel-decode improvement from the perf issue lands first).
3. Re-run 06_taxonomy + retrain; spot-check the confidence thresholds still behave (the -unsure boundary may shift).
4. Update the ML smoke test tag and its cache key.

## Acceptance

- No QuickGELU warning at model load.
- `tests/test_ml_smoke.py` green with the new tag.
- Re-embedded library + regenerated assignments verified in the review UI.

**Comment by yves-vogl** [comment]
## Analysis + bounded experiment (2026-08-27): the divergence is REAL — keeping this open for a dedicated migration pass

Ran the bounded experiment instead of the migration: loaded the same OpenAI checkpoint through both configs (`ViT-B-32` = production today, GELU activations, warns; `ViT-B-32-quickgelu` = the config the weights were trained with, no warning), embedded a deterministic 1,024-photo sample of the real library with both, and quantified the divergence. Setup: Apple M-series/MPS, exact venv pins from requirements.txt, taxonomy prompts extracted from `06_taxonomy.py` via AST and encoded with the matching model variant, softmax/LOGIT_SCALE convention identical to production.

### Sanity

- Production `embeddings/clip_vitb32.npy` matches the plain `ViT-B-32` space exactly (mean cos **1.0000** on all 1,024 sampled uuids); mean cos to the quickgelu space is **0.9718** — the two spaces are close but not interchangeable.

### 1. Similarity-ranking divergence (the `find_similar` feature)

| metric | value |
|---|---|
| top-10 neighbour overlap | mean **0.745**, median 0.800, p5 0.500, min 0.100 |
| top-20 neighbour overlap | mean **0.782**, median 0.800, p5 0.558, min 0.200 |
| top-1 neighbour identical | **64.8 %** |
| Spearman rho, full per-query ranking | mean **0.951**, p5 0.918, min 0.725 |

→ **~35 % of queries get a different best match**, and a quarter of every top-10 list changes. Not negligible.

### 2. Zero-shot taxonomy divergence (06_taxonomy)

| metric | value |
|---|---|
| top-1 label agreement | **84.9 %** (155/1024 photos change label) |
| confidence-band migrations (0.80 / 0.50 boundaries) | **149/1024 = 14.6 %** |
| confidence delta (quickgelu − plain) | mean **+0.022**, p95 \|Δ\| 0.256 |
| photos ≥ CONF_MAIN 0.80 | 256 → **286** |
| photos < CONF_UNSURE 0.50 | 602 → **564** |

→ The quickgelu (correct) variant is systematically **more confident** — consistent with the open_clip README's documented accuracy drop for the mismatched config. The 0.80/0.50 thresholds were tuned against the degraded space and will need a spot-check after migration.

### Migration plan (dedicated pass — do NOT do partially)

1. **Switch all SIX call sites in one change** (the issue text said four; the full list is): `01_embed.py`, `03_classify_cluster.py` (+tokenizer), `06_taxonomy.py` (+tokenizer), `07_incremental.py` (line ~183), `embed_test.py`, `tests/test_ml_smoke.py` (+tokenizer). A partial switch mixes incompatible spaces silently.
2. **Guard against silent space-mixing first**: `01_embed.py`'s resume logic would happily *append* quickgelu embeddings to the existing plain-GELU `clip_vitb32.npy` (there is no model tag stored anywhere in the embedding artefacts). Before re-embedding, either write a sidecar tag file checked by 01/07 or use a new filename (e.g. `clip_vitb32_quickgelu.npy`) and update the readers (server.py, 03, 06_retrain, 06_taxonomy, 07). Recommendation: sidecar tag + hard fail on mismatch — cheapest and permanent.
3. **Full re-embed**: measured cost after #39 (parallel loader): **~9.5 min** for 81.5k photos at ~144 img/s on this box (was ~15.5 min serial). Delete `clip_vitb32.npy` + `clip_vitb32_uuids.json`, run `01_embed.py` cold.
4. **Re-run 06_taxonomy + 06_retrain**; expect ≈15 % of zero-shot labels to move and the confidence distribution to shift up (+30 photos over 0.80, −38 under 0.50 per 1,024 in the sample). Spot-check CONF_MAIN=0.80 / CONF_UNSURE=0.50 — the `-unsure` boundary shifts.
5. **Corrections are unaffected** (uuid-keyed, no embedding dependency); the retrain classifier must be retrained on the new space in the same pass.
6. **CI**: update the model tag in `tests/test_ml_smoke.py`; the HF cache key in `ml-smoke.yml` can stay (the quickgelu config downloads the *same* openai checkpoint — only the activation function in the model config changes), rename the key only for clarity.
7. Verify in the review UI (find-similar sanity + album spot-check) before deleting any backup of the old `.npy`.

Ordering note from the issue confirmed: #39 landed first, so the re-embed is now ~40 % cheaper than at filing time.

**Comment by yves-vogl** [comment]
## Migration executed 2026-08-27 — all 7 plan steps complete ✅

**PRs (both merged green):** #50 (sidecar model-tag guard, steps 1+2 prerequisite) → #52 (QuickGELU config switch + CI revert guards, closes this issue). #51 was the same content as #52, auto-closed when its stacked base branch was deleted.

### Per plan step

1. **All six call sites switched atomically** — they now load through `model_tag.MODEL_NAME` (single constant), so no call site can ever drift to a different config again.
2. **Resume-append trap closed first** (#50): sidecar tag `embeddings/clip_vitb32.model.json`, hard fail on mismatch AND on untagged legacy artefacts (no silent adoption). Pinned by fast-lane unit tests + an ml-smoke e2e test (tampered sidecar → `01_embed.py` exits non-zero, artefacts byte-identical).
3. **Full re-embed:** 81,555/81,555 photos, **0 failures**, 805.7 s (13.4 min) at 101.2 img/s on MPS with the #39 loader. Pre-flight discovery: 22,264 of the June derivative paths had gone stale (Photos.app rebuilt its derivatives cache); all were refreshed from PhotosDB with the uuid set unchanged (`inventory.csv.pre-quickgelu.bak` kept).
4. **06_taxonomy + 06_retrain re-run** on the new space. The quickgelu space is markedly more confident, as the bounded experiment predicted: assignments moved from 33,649 main / 14,092 unsure / 33,814 Diverses to **39,821 main / 21,074 unsure / 20,592 Diverses** (Diverses −39 %). Retrain: 68 corrections → 67 strong labels (1 below MIN_CORR) + 36,315 weak, 22 classes.
5. **Corrections replay validated:** 68/68 final non-skip corrections land in the new `assignments.csv` exactly as corrected (album match, conf 1.0, `source=correction`) via the shared `corrections_log` replay from #48.
6. **CI:** fast lane now pins `MODEL_NAME == "ViT-B-32-quickgelu"` on every PR (no torch needed); ml-smoke asserts the model loads with **zero QuickGELU warnings**; HF cache key renamed for clarity.
7. **Verification on the fixed 1,024-photo corpus** (seed-42 deterministic sample, fresh in-process quickgelu reference embed):
   - **Self-consistency: production space vs quickgelu reference — mean cos 1.0000, min 1.0000.** The old space scored 0.9688 against the same reference.
   - uuid set AND order identical to pre-migration (81,555, 0 missing, 0 added).
   - Old-vs-new divergence confirms the original analysis: mean cos 0.9688, top-10 neighbour overlap 0.742 (analysis: 0.745), top-1 neighbour identical 61.5 % (analysis: 64.8 %).
   - Review-UI backend: contact sheets regenerated (46 albums), `/api/similar` spot-checked through the real server — ranked descending, query excluded, semantically coherent neighbours (screenshots → screenshots, motorcycles → motorcycles).

**Backups kept** until Yves has eyeballed the review UI: `embeddings/*.pre-quickgelu.bak`, `metadata/assignments.csv.pre-quickgelu.bak`, `metadata/multi_assignments.csv.pre-quickgelu.bak`, `metadata/inventory.csv.pre-quickgelu.bak` (all gitignored via the new `*.bak` rules).
